### ISSUE TITLE
Basecamp Library and associated ongoing Study mission

### ISSUE THREAD

**Issue by descan** [CleverRaven/Cataclysm-DDA#59723]
### Is your feature request related to a problem? Please describe.

Having to handhold all my companions through learning and studying and upping their skills. They're adults, they know how to read damnit. I shouldn't have to shove the book in their hands and order them to do it, I should be able to give them a standing order to get better! We're trying to survive! Gotta have skills!

### Solution you would like.

I would love to be able to build a library in my basecamp, stock it with books, and give my companions an ongoing priority for studying and raising their skills.

### Describe alternatives you have considered.

Have companions automatically grab books to learn when they're not doing any activity. Personally I would prefer having to build a study area, and being able to restrict which books they can use (i.e. not grab the book I need for a bunch of recipes unless I've specifically put it in the library...)

### Additional context

this may already be in the game? it's a bit hard to parse the basecamp code, and I haven't encountered it in-game yet.

**Comment by PatrikLundell** [comment]
I don't think this needs to be restricted to base camp use, but can rather be a zone, which would make it available to those who don't want to deal with base camps. Since zones can be tied to vehicle tiles, such an implementation would make it possible for nomadic characters to have library tiles in their mobile bases.
(Hm, mobile base camps?).

Note that this doesn't mean you couldn't have a base camp blueprint for a library, but that would be more for flavor.

However, none of this currently exists.
There are a few technical problems that would have to be dealt with, of varying difficulties for this to work:
- A way to assign background activities to companions, so whenever they're done with a task they take up the background one (or "self maintenance" ones, such as eat/sleep). Note that you'd probably want to at least have the option to get a clear notification (interrupt popup?) that their assigned tasks are done, as the current "X is finished with the Y activity" note in the message log isn't exactly standing out, and my cue is typically that the companions are standing around near the player.
- A way for the game to "catch up" with background activities that should have been performed while the player gets the companions back into the reality bubble. The only such activity I know of that's implemented this is vehicle dismantling (where the companion drops all the parts that should have been removed during the absence as the companion gets back into the bubble). This gets tricky with background activities as those might involve moving to a location (the "library" that's not yet in the reality bubble when it touches the companion. If we (at least initially) restrict the reality bubble return logic to only catch up on background tasks for which the means are at hand immediately (like catching up on the reading of a book in the inventory) it might be easier to get started.
- A way for companions to be better at returning "loans", so they don't accumulate books in their inventories, but return them to any library zone when done.
- Modify reading to continue until the learning potential of a book has been exhausted, rather than until the next level.
- Some way to specify what the companions should study (so they're not reading at random), as well as what to do when no progress is available (no books for that currently available in the library). Possibly some kind of study priority order, so the top priority is returned to when back to the library and a book for the top subject becomes available.
- Similar functionality for practice recipes. That's more complicated, as practice recipes tend to use resources you may not want to be spent without oversight. As far as I understand, companion crafting doesn't apply proficiencies (properly?) currently, but they should.

There is one key advantage to the base camp functionality when it comes to implementation, and that is that companions are send off into the void, so the reality bubble issues are bypassed. However, that's offset both by it being restricted to base camps, and by the need for the player to actively recall companions from the void after their activities are done.

The rambling above has likely missed a number of relevant points, and may well miss the mark on others...

**Comment by descan** [comment]
You're quite right, actually - it needed be restricted to basecamps. (I've just been fiddling with the last few days to check them out, so they've been on my mind.) If one goes that route, maybe we could get a Bookcase vehicle part too? ;P

Aren't the background tasks (supposed to be) handle by the task priority lists when assigning workers to a basecamp? That's part of why I suggested it working with the basecamp system. That priority list isn't perfect, however (they don't seem to follow it all the time, sometimes hanging around and doing nothing while there's vehicles to deconstruct f. ex., and you have to do everyones priority list individually - you can't copy from another worker, or better yet, have a set of priority lists labelled for Guards, Smiths, Farmers, Generalists, etc.)

The rest of your message all makes sense to me. Although with the last one, I would say similarly to having a Library zone where they use only the books there and not just scrounging them from wherever, a Practice zone where you can put your surplus for them to practice with and not eat into your vital supplies might do the trick?

**Comment by PatrikLundell** [comment]
Breaks when studying are good, but failing exams is bad, so I hope you're balancing it reasonably.

I've never assigned anyone to base camps, except by accident, which is then immediately reversed...

I just made a quick experiment of assigning companions to the camp, and it didn't work out well...
I sent one to primarily do construction, and she ran into a room to get stuck there, presumably trying to each a construction site on the other side of the wall. A second moron was primarily tasked with dismantling vehicles, but for some reason went after the first one to get stuck in the same room, rather than moving in a direction 90 degrees off of that to actually reach the vehicle being deconstructed (and if he went for tools, the warehouse is in the 180 degrees direction compared to the one he headed in).

So, you're right in that there's a system that's intended to handle background activities, but it seems to be broken. It may well be that the logic of that system isn't what's broken, but rather companion path finding, as I've had to guide companions rather carefully to get them to construct things, or they run into walls and just get stuck there, either tying to get parts or trying to get to the site, so I've hauled all the construction material to a tile next to the construction site, painted a Supplies zone on top of it, and guided the moron to stand on it (after supplying it with the needed tools), before ordering construction (and in some cases I had to disable other construction zones as the idiot tried to fail to go to those instead). I've also had issues with sorting out things where they, again, just got stuck apparently trying to path through a wall. I worked around that by removing the Unsorted zone on the other side...

Assuming the background task system was actually working, it does indeed look like it could be built upon, but there's probably need for a significant UI change to allow you to specify study as a task with some priority, but then expand that activity to further detail what to study.
It doesn't look like the base camp background activity orders contain anything that's base camp specific, though, so it ought to be possible to implement that via dialog rather than (or in addition to) the bulletin board.

I'm not sure how useful set or copied task lists would be. Setting more than one companion to dismantle vehicles can crash the game, and for study you'd be better off trying to get different companions trying to learn different skills, rather than all competing over the same books and get the same profile (so everyone is capable of batch crafting some things with you, but nobody can help you with others).
Setting up activities would be a hassle only if you have a large group of companions, and if you do, you're probably going to struggle to keep the all occupied.

A practice supplies zone may be useful, but it would be a hassle to keep it supplied with what's needed for the moment.

**Comment by descan** [comment]
I don't think it was a pathing issue, I think it was a needs/priority issue. When my companions do that thing - run into a building and don't do work that's already laid out and they're prioritized, what they end up doing is sitting on the nearest Comfortable Place (A bed, a chair, even a table at one point). When I talk to them and order them specifically to do something, they'll be more likely to go do it (Same as when I myself move out and can see the construction/dissassembly work that's needing to be done - the workers are more likely to come out when I'm looking at the vehicles than when I'm off crafting)

However, when I ask them what activity they're working on, they're not saying "I'm trying to construct," they're "Not doing much of anything." As well, I'm not getting error-lag from lack of tools or other errors indicating they're even *attempting* to do the work. Whatever it is that's stopping them is either a bug, or an esoteric system interaction that should need be splayed open a bit more for players to understand why their workers aren't working and just lazing about. It could be both, even - if I have "tidy up" and "sort out loot" (Which are, for some reason, two different priorities and I don't even know what "tidy up" does) at even 1 priority, they'll always try to do that over their higher-order priorities. (constantly, over and over, even when my unsorted zones are disabled - I have to delete them to get them to stop, or disable their priorities, or else I get "X has finished sorting out all available loot!"). 

So yeah, that priority list is a gamble, and needs to be taken a look at - either fixed, deleted, or moved out of basecamp (and then fixed. or deleted.)

**Comment by PatrikLundell** [comment]
Yes, it was probably something else looking like the pathing problems I've seen, as the companion claimed to guard the position, when that wasn't anything I'd given a value. Whether priority 1 is higher or lower than 2 is something I don't know for certain, but when I checked construction was the only activity assigned (when I tried the direct order to construct the companion disappeared looking for tools, giving up shortly thereafter).

I suspect "tidy up" doesn't do anything, and may have been disconnected at some point in the past. The reason for that guess is that I've fixed a weird "tidying up" activity that vehicle dismantlers used to engage in when their activity was broken off e.g. by darkness, but not by finishing the task completely. In that case the bugger started to dump all sorts of crap into Unsorted zones, including those of my vehicles. My fix was to remove that buggy and unexpected activity from the tail of vehicle dismantling, but it remains tied to fishing, because I don't know why it's there in the first place, and nobody provided any info, so I made the minimum change possible. I believe the function wasn't tied to the faction camp tidy up operation, though, but I may misremember.

Apart from seemingly not working properly, I'm not comfortable with unknown behavior kicking in (such as the potential risk that somebody would decide to attack and dismantle a handjack "vehicle" because it happened to be pulled through a dismantling zone), and I also want to be notified when they're done so I can give them their next task.

Ideally, the fixing part should be done by someone who understands how the current system is supposed to know and wants to use it, or, failing that, someone who fulfills one of these criteria.

**Comment by chrispikula** [comment]
> * A way for companions to be better at returning "loans", so they don't accumulate books in their inventories, but return them to any library zone when done.

I'd go one step further, and say don't even simulate them picking up the books.  Maybe give them phantom copies of the books they want to read, with transformation times to nothing of 1 tick?  As soon as they are back in the reality bubble, they go away.

> * Some way to specify what the companions should study (so they're not reading at random), as well as what to do when no progress is available (no books for that currently available in the library). Possibly some kind of study priority order, so the top priority is returned to when back to the library and a book for the top subject becomes available.

Likewise, something to memorize recipes (specific ones?) instead of just theoretical skill levels.

**Comment by PatrikLundell** [comment]
The problem with not actually remove the books from the Library is that several companions can study the same book at different levels at the same time. If locking books while in use with an "in progress" job in the companion's inventory is easier to implement than physically removing it that would work.
However, based on my brief fiddling with basecamp background activities it doesn't seem those are abstracted, and non base camp library zone usage would likely need to actually take the books, so I suspect a take loan and return functionality may be needed. As a bonus, it might actually work with tools as well.

Companion memorization of recipes is currently useless, as batch assistance depends on player recipe selection and base camp crafting is dependent on cludgy base camp recipe selection lists. The future marginal use case for memorized recipes would be when a book containing a recipe is currently in use.
Companion learning of proficiencies (through crafting, world actions (if any) and practice recipes) would definitely be needed, though, when companion crafting makes use of proficiencies (I don't think that logic has been introduced to companion crafting).
